kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/jms/BlockingResponseCallback.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+package org.kie.server.client.jms;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+
+import org.kie.server.api.marshalling.Marshaller;
+import org.kie.server.api.model.ServiceResponse;
+import org.kie.server.api.model.ServiceResponsesList;
+import org.kie.server.api.model.Wrapped;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Simple blocking response callback backed by blocking queue that will allow sequential access to
+ * responses and block
+ * <ul>
+ *     <li>client if there is no message yet</li>
+ *     <li>server if the queue is full</li>
+ * </ul>.
+ */
+public class BlockingResponseCallback implements ResponseCallback {
+
+    private static final Logger logger = LoggerFactory.getLogger(BlockingResponseCallback.class);
+    private BlockingQueue<ServiceResponsesList> responses = new ArrayBlockingQueue<ServiceResponsesList>(10);
+
+    private Marshaller marshaller;
+
+    public BlockingResponseCallback(Marshaller marshaller) {
```

Comment:
no, we don't need always the marshaller in case you want to get raw ServiceResponseList and then deal with reading the response yourself. If you would like to have this done for you then you need to provide your marshaller that is container specific, so it needs to be provided by the one how registers the callback.